puppet/lib/puppet/parser/functions/query.rb

```diff
@@ -0,0 +1,45 @@
+Puppet::Parser::Functions.newfunction(:query, :type => :rvalue, :arity => -3, :doc => <<-EOT
+Query PuppetDB
+
+Usage:
+  query(["from", "resources",
+         ["and", ["=", "type", "Class"],
+          ["=", "title", "Puppet_enterprise::Profile::Puppetdb"],
+          ["not", ["=", "certname", $certname]]]])
+  [{"certname" => "foo.bar.com"
+    "environment" => "production",
+    "parameters" => {"master_certname" => "master.foo.lan"
+                     ...}} ...]
+
+  query(["from", "resources",
+         ["and", ["=", "type", "Class"],
+          ["=", "title", "Puppet_enterprise::Profile::Puppetdb"],
+          ["not", ["=", "certname", $certname]]]],
+        'certname')
+  {"foo.bar.com" => {"certname" => "foo.bar.com"
+                     "parameters" => {"master_certname" => "master.foo.lan"
+                                      ...}}}
+
+  query(["from", "resources",
+         ["and", ["=", "type", "Class"],
+          ["=", "title", "Puppet_enterprise::Profile::Puppetdb"],
+          ["not", ["=", "certname", $certname]]]]'
+        'certname',
+        true)
+  ["foo.bar.com", "baz.buzz.net", ...]
+EOT
+                                     ) do |args|
```

Comment:
I think a unit test, that mocks the final call would be a good thing here, anything that at least checks the code syntax would be nice, right now without any unit test even basic ruby validation doesn't occur I think.